### ISSUE TITLE
bundle fails with "Module c not present"

### ISSUE THREAD

**Issue by andrenarchy** [jspm/jspm-cli#1491]
In [paperhive-frontend](https://github.com/paperhive/paperhive-frontend/tree/322fd7ae7d80dba5c0be7aabbf63045019feab49) everything runs smoothly without bundling (jspm 0.17.0-beta.5). After creating a bundle with 
```
jspm build app/index.ts build/index.js
```
the browser fails while executing the bundle:
```
Uncaught Module c not present.
```
It seems to fail in line 2 of the following generated code:
```
$__System.registerDynamic("b", ["c"], true, function($__require, exports, module) {
  var process = $__require("c");
  var global = this,
      __define = global.define;
  global.define = undefined;
  (function(global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() : typeof define === 'function' && define.amd ? define(factory) : global.moment = factory();
  }(this, function() {
    'use strict';
    var hookCallback;
    function utils_hooks__hooks() {
      return hookCallback.apply(null, arguments);
    }
[...]
```

Any hints for getting a working bundle are highly appreciated! :smiley: Let me know if I can help to track this down. I'm a bit clueless with this generated code...

**Comment by andrenarchy** [comment]
I just found out that using a standart (i.e., non-self-executing) bundle created with
```
jspm bundle app/index.ts build/bundle.js
```
works as expected. 

I don't know if it helps to track down this issue but I compared the standard bundle code with the self-executing bundle code and the dependency `c` in the above output should be the `process` module (`github:jspm/nodelibs-process@0.2.0-alpha`). In the standard bundle it's correctly resolved but it fails in the self-executing bundle.

@guybedford is this the right repo or should I re-open the issue in https://github.com/systemjs/builder?

**Comment by guybedford** [comment]
@andrenarchy it looks like you are chaining the TypeScript plugin with Babel. For this case it is advisable to set `typescriptOptions: { modules: 'es6' }` (I think it's that?), because otherwise the dependencies won't be normalized correctly.

**Comment by guybedford** [comment]
(Effectively having Babel process System.register directly stops the Babel transpiler from being able to handle dependency normalization)

**Comment by andrenarchy** [comment]
I've set `module: 'es6'` in [tsconfig.json](https://github.com/paperhive/paperhive-frontend/blob/jspm/tsconfig.json) and typescript uses this file via `tsconfig: true` in [jspm.config.js](https://github.com/paperhive/paperhive-frontend/blob/jspm/jspm.config.js#L9). Do you really mean `modules` (plural)?

**Comment by andrenarchy** [comment]
Adding `modules: 'es6'` to `jspm.config.js` doesn't fix the problem.

**Comment by guybedford** [comment]
I just came across this myself. I believe it is fixed in https://github.com/systemjs/builder/commit/0c8e88d3596e37c739d713d45b725b37e871eb96, will post a release later tomorrow.

**Comment by andrenarchy** [comment]
This would be super awesome and I'd totally buy you a beer/coffee/whatever. :smile:

**Comment by davis** [comment]
hm, so i followed this guide http://jspm.io/0.17-beta-guide/bundling.html
and at this step i tried doing a jspm build, but i run into a `Uncaught Module 4 not present.`

i updated the builder code with that code in the commit, but i'm still running into the issue

**Comment by guybedford** [comment]
@davis it's definitely working on edge. Try it again when the release is out.

**Comment by davis** [comment]
ah, it is indeed working!! thank you!

**Comment by andrenarchy** [comment]
With jspm 0.17-beta.6 I get `Module lodash not present.`

**Comment by guybedford** [comment]
@andrenarchy did you still follow https://github.com/jspm/jspm-cli/issues/1491#issuecomment-179430694? Any remaining issues have to do entirely with chaining the plugin with the transpiler. Transpiling `System.register` will not catch dependencies.

**Comment by andrenarchy** [comment]
Yes, I tried `typescriptOptions: { modules: 'es6' }` as well but it looks like `typescriptOptions: { module: 'es6' }` is the correct option and I've always had this. Here are the [json.config.js](https://github.com/paperhive/paperhive-frontend/blob/c5c3b19ba485727728e6f5d52c0fb1f12bfc0fd5/jspm.config.js) and [tsconfig.json](https://github.com/paperhive/paperhive-frontend/blob/c5c3b19ba485727728e6f5d52c0fb1f12bfc0fd5/tsconfig.json) files that I use.

@guybedford What exactly do you mean with "Transpiling System.register will not catch dependencies."?

**Comment by guybedford** [comment]
`typescriptOptions: { module: 'es6' }` doesn't seem to be in your jspm.config.js though? The point is to verify for certain that what the TypeScript plugin is spitting out is ES6 code and not System.register code. I can't really offer more support than that unless you can confirm for a fact that this is happening on ES6 code through the transpiler with a simplified test case that verifies it.

**Comment by andrenarchy** [comment]
I use `tsconfig: true` in order to include typescript-specific options from [tsconfig.json](https://github.com/paperhive/paperhive-frontend/blob/c5c3b19ba485727728e6f5d52c0fb1f12bfc0fd5/tsconfig.json) (this way also my editor knows about my ts config). Adding `module: "es6"` directly also leads to `Module lodash not present`.

How can I verify that the ts plugin generates es6 code?

**Comment by guybedford** [comment]
@andrenarchy the first step debugging here is to work out what name is missing. You can do this via `jspm build app/index.ts --skip-encode-names`. Then that should help to understand further where it is missing the dependency.

**Comment by guybedford** [comment]
If you can share more info here I'm happy to help.

**Comment by andrenarchy** [comment]
I just want to share that all issues are fixed in the most recent versions of jspm (0.17.0-beta.11) and the typescript+babel plugins. @guybedford thanks for your work and help! :+1:
